feat: Governance topic approval via Terraform and GitHub Actions (#776)
aiven/aiven-docs: docs/products/kafka/howto/terraform-governance-approvals.md

---
title: Manage approvals in Aiven for Apache Kafka® Governance using Terraform & GitHub
sidebar_label: Governance with Terraform
---

import TerraformSample from '@site/src/components/CodeSamples/TerraformSample';

Aiven for Apache Kafka® Governance lets you manage approval workflows for Apache Kafka topic changes using Terraform and GitHub Actions.

## Prerequisites

> - [Governance](/docs/products/kafka/howto/enable-governance) enabled for your organization

Reviewer: Governance does not actually have to be enabled at all. In fact, no interaction with console or aiven-core is required. The checker does not know anything about aiven-core.

Reviewer 2: But once they execute terraform apply for topics at least it will do some additional checks at the API level and also stops people making changes without authorization.

Reviewer: Hmm, what is actually the difference of enabling Terraform governance and not having governance enabled? 🤔

   - Ensure that only mapped users can request or approve changes.

1. Submit and validate changes in GitHub

   - A user submits a pull request (PR) to modify an Apache Kafka topic.
   - The GitHub Action automatically runs governance compliance checks.
   - If the requester lacks approval permissions, the check fails.

1. Approve and apply changes

   - An authorized user approves the PR.
   - The GitHub Action reruns governance validation.
   - If all policies are met, the PR passes, and Terraform applies the changes.

### Approval rules

Governance enforces approval policies to maintain security and compliance:

- If an approver is a member of multiple teams, their approval applies to all teams
  they belong to.
- If a second approval is required, another member of the same user group must approve.
- The pull request creator must be a member of the user group that owns the Apache
  Kafka topic.

## Set up governance approvals

Set up approval workflows using Terraform and GitHub Actions.

### Step 1. Define topic ownership and user groups

To restrict modifications to a specific user group, specify the

> `owner_user_group_id` in your Terraform configuration:

Reviewer 2: This is mandatory when governance is enabled

   runs-on: ubuntu-latest
   steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Get PR approvers
        id: get_approvers
        uses: octokit/request-action@v2.x
        with:
          route: GET /repos/${{ github.repository }}/pulls/${{ github.event.pull_request.number }}/reviews
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

      - name: Run compliance check
        uses: aiven/aiven-terraform-governance-compliance-checker@v0.1.0
        with:
          requester: ${{ github.event.pull_request.user.login }}
          approvers: ${{ steps.get_approvers.outputs.data }}
          aiven_token: ${{ secrets.AIVEN_API_TOKEN }}
```

### Step 4. Validate and apply changes using Terraform

Before applying approved changes, check governance compliance with Terraform:

```bash
terraform plan -out=tfplan
terraform show -json tfplan > plan.json
```

Use `plan.json` for governance compliance checks before merging the pull request.
If your workflow does not automatically apply Terraform changes, run:

```bash
terraform apply "tfplan"
```

Only approved changes are applied. Terraform fails if changes violate governance rules.

> 

Author: update the line here.

### Step 5. View compliance check results

After validation, check the GitHub Actions logs or Terraform output to verify compliance.

- Successful check: The request meets governance requirements.

  ```json
  {
    "ok": true
  }
  ```

- Failed check: The request does not meet governance rules, and an error message
  appears in the pull request.

  ```json
  {
    "ok": false,
    "errors": [
    {
      "error": "Requesting user is not a member of the owner group",
      "resource": "aiven_kafka_topic.orders"
    }
   ]
  }
  ```

> If a check fails, GitHub blocks the pull request until the issue is resolved. To fix

Reviewer: Same as above, the GitHub action can't interact with the repository configurations so _we_ can't block it. It is up to the repository maintainers to configure the repository so, that the Worklfow using this check is required to pass. 

Also, if the GitHub action output has errors, it does not make the Workflow fails. This power is again in the hands of the repository maintainers.

In summary; we are limited to telling you if the changes that would be applied comply or not with aiven governance rules. It is up to the user to enforce this whatever the environment they are using.

Author: update this line.

---
title: Manage approvals in Aiven for Apache Kafka® Governance using Terraform & GitHub
sidebar_label: Governance with Terraform
---

import RelatedPages from "@site/src/components/RelatedPages";
import TerraformSample from '@site/src/components/CodeSamples/TerraformSample';

Aiven for Apache Kafka® Governance lets you manage approval workflows for Apache Kafka topic changes using Terraform and GitHub Actions.

To manage approvals in the Aiven Console, see
[Manage approvals in the Aiven Console](/docs/products/kafka/howto/approvals).

## Prerequisites

> - [Authentication token](/docs/platform/howto/create_authentication_token) for an

Reviewer 3: Small fix to the link - tokens for application users are created in a different place than the personal tokens. I also suggested changing this to "tokens" since we have moved to naming these "personal tokens" for users and "application tokens" for application users. You could also use "application token" here.

```suggestion
- [Token](/docs/platform/howto/manage-application-users) for an
```

changes using Terraform and GitHub Actions. These workflows follow three key principles:

- **Topic ownership**: Apache Kafka topics are assigned to user groups using
`owner_user_group_id`. Only members of the assigned group can approve changes to
that topic.
- **External identity mapping**: The `aiven_external_identity` resource links GitHub
usernames to Aiven users, ensuring that requesters and approvers are correctly identified.
- **Approval validation**: A GitHub Action automatically checks that pull request
creators and approvers meet governance requirements before changes are applied.

### Workflow steps:

1. Define governance policies using Aiven Terraform Provider:

   - Assign topic ownership (control who can approve changes) using `owner_user_group_id`.
   - Configure user groups (`aiven_organization_user_group`) to manage approval
     permissions.
   - Enforce team-based approval policies to ensure compliance.
   - Automatically validate changes against governance policies.

1. Map GitHub users to Aiven identities:

   - Link GitHub usernames to Aiven users using the `aiven_external_identity` resource.
   - Ensure that only mapped users can request or approve changes.

> 1. Submit and validate changes in GitHub

Reviewer 3: ```suggestion
1. Submit and validate changes in GitHub:
```

   - A user submits a pull request (PR) to modify an Apache Kafka topic.
   - The GitHub Action automatically runs governance compliance checks.

>    - If the requester lacks approval permissions, the check fails.

Reviewer 2: ```suggestion
   - Without all required approvals, the check fails.
```
This should always fail basically when you first create it. I think this makes it slightly clearer?

> 1. Approve and apply changes

Reviewer 3: ```suggestion
1. Approve and apply changes:
```

   - An authorized user approves the PR.
   - The GitHub Action reruns governance validation.
   - If all policies are met, the PR passes, and Terraform applies the changes.

### Approval rules

Governance enforces approval policies to maintain security and compliance:

- If an approver is a member of multiple teams, their approval applies to all teams
  they belong to.

> - If a second approval is required, another member of the same user group must approve.

Reviewer 2: ```suggestion
- If a second approval is required, another member of the same user group must approve.
```
Is it a bit confusing that we switch between talking about user groups and teams?

Author: Yeah, I agree. I think it is better to use user groups instead of teams.

- The pull request creator must be a member of the user group that owns the Apache
  Kafka topic.

## Set up governance approvals

Set up approval workflows using Terraform and GitHub Actions.

### Step 1. Define topic ownership and user groups

To restrict modifications to authorized users, you must specify `owner_user_group_id`
in your Terraform configuration:

<TerraformSample filename="resources/aiven_kafka_topic/resource.tf" />

Setting `owner_user_group_id` alone does not enforce approvals. The GitHub Action must
be integrated to validate changes and enforce compliance.

### Step 2. Map GitHub users to Aiven identities

To verify requesters and approvers, assign them to the appropriate user group
using the `aiven_external_identity` resource. For more information, see the
[Aiven Terraform Provider documentation](https://registry.terraform.io/providers/aiven/aiven/latest/docs/data-sources/external_identity).

**Example Terraform configuration**

> ```hcl

Reviewer 3: This whole section seems to me a good use case for inserting the example usage from the TF repo to avoid maintaining these in two places and these going out of date. I see there's no example usage for the data source, yet. I could open a PR for that today.

Author: Sounds good! Let me know once that example is in, and I’ll update the docs to point to it. I can also do that in a separate PR if needed; just trying to get this out ASAP.

Reviewer 3: Totally understand that! I'll do a proper review of this after my meetings this morning and then you can just get this merged and I'll let you know when things are ready on the TF side later for updates. It usually takes a few days to get doc updates merged in the TF repo, so don't want to hold you up for that. :)

  organization_id       = data.aiven_organization.example_org.id
  internal_user_id      = "aiven_user_id"
  external_user_id      = "github_username"
  external_service_name = "github"
}
```

- This mapping ensures that GitHub users are recognized as Aiven users for governance
  approvals.
- Only mapped users can request or approve changes.
- Repeat this for all users who need to request or approve changes.

### Step 3. Set up compliance checks in GitHub

In your GitHub repository, define a GitHub Actions workflow to enforce governance
approvals.

**Example GitHub Actions workflow**

```yaml
 name: Kafka governance compliance check

 on:
   pull_request:
     types: [opened, synchronize, reopened]
   pull_request_review:
     types: [submitted]

 jobs:
   compliance:
   runs-on: ubuntu-latest
   steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Get PR approvers
        id: get_approvers
        uses: octokit/request-action@v2.x
        with:
          route: GET /repos/${{ github.repository }}/pulls/${{ github.event.pull_request.number }}/reviews
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

      - name: Run compliance check

>         uses: aiven/aiven-terraform-governance-compliance-checker@v0.1.0

Reviewer: The repo name has updated to:
Aiven-Open/Aiven-apache-kafka-governance-compliance-checker-for-terraform@v1.2.0

Author: Thanks! Update this.
